// @sap-ux/preview-middleware/dist/client/i18n.js
"use strict";

sap.ui.define(["sap/base/i18n/ResourceBundle"], function (ResourceBundle) {
  "use strict";

  const BUNDLE_CACHE = {};
  async function getResourceBundle(key) {
    const cachedBundle = BUNDLE_CACHE[key];
    if (cachedBundle) {
      return cachedBundle;
    }
    const bundle = await ResourceBundle.create({
      bundleUrl: '/preview/client/messagebundle.properties',
      url: '/preview/client/messagebundle.properties',
      supportedLocales: [''],
      locale: '',
      async: true
    });
    BUNDLE_CACHE[key] = bundle;
    return bundle;
  }
  class TextBundle {
    constructor(bundle) {
      this.bundle = bundle;
    }
    getText(key, args) {
      return this.bundle.getText(key, args) ?? key;
    }
  }
  async function getTextBundle() {
    let key = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'open.ux.preview.client';
    const bundle = await getResourceBundle(key);
    return new TextBundle(bundle);
  }
  var __exports = {
    __esModule: true
  };
  __exports.getResourceBundle = getResourceBundle;
  __exports.TextBundle = TextBundle;
  __exports.getTextBundle = getTextBundle;
  return __exports;
});
