// @sap-ux/preview-middleware/dist/client/adp/quick-actions/fe-v2/lr-toggle-clear-filter-bar.js
'use strict';
sap.ui.define([
    'sap/ui/rta/command/CommandFactory',
    'open/ux/preview/client/thirdparty/@sap-ux-private/control-property-editor-common',
    '../../../cpe/quick-actions/utils',
    '../../../utils/core'
], function (CommandFactory, ___sap_ux_private_control_property_editor_common, _____cpe_quick_actions_utils, _____utils_core) {
    'use strict';
    const SIMPLE_QUICK_ACTION_KIND = ___sap_ux_private_control_property_editor_common['SIMPLE_QUICK_ACTION_KIND'];
    const pageHasControlId = _____cpe_quick_actions_utils['pageHasControlId'];
    const getControlById = _____utils_core['getControlById'];
    const ENABLE_CLEAR_FILTER_BAR_TYPE = 'enable-clear-filter-bar';
    const PROPERTY_NAME = 'showClearOnFB';
    const CONTROL_TYPE = 'sap.ui.comp.smartfilterbar.SmartFilterBar';
    class ToggleClearFilterBarQuickAction {
        kind = SIMPLE_QUICK_ACTION_KIND;
        type = ENABLE_CLEAR_FILTER_BAR_TYPE;
        get id() {
            return `${ this.context.key }-${ this.type }`;
        }
        get isActive() {
            return !!this.filterBar;
        }
        isClearButtonEnabled = false;
        constructor(context) {
            this.context = context;
        }
        initialize() {
            const controls = this.context.controlIndex[CONTROL_TYPE] ?? [];
            for (const control of controls) {
                const isActionApplicable = pageHasControlId(this.context.view, control.controlId);
                const modifiedControl = getControlById(control.controlId);
                if (isActionApplicable && modifiedControl) {
                    this.isClearButtonEnabled = modifiedControl.getShowClearOnFB();
                    this.filterBar = modifiedControl;
                }
            }
        }
        getActionObject() {
            const key = this.isClearButtonEnabled ? 'V2_QUICK_ACTION_LR_DISABLE_CLEAR_FILTER_BAR' : 'V2_QUICK_ACTION_LR_ENABLE_CLEAR_FILTER_BAR';
            return {
                kind: SIMPLE_QUICK_ACTION_KIND,
                id: this.id,
                enabled: this.isActive,
                title: this.context.resourceBundle.getText(key)
            };
        }
        async execute() {
            if (this.filterBar) {
                const {flexSettings} = this.context;
                const modifiedValue = {
                    generator: flexSettings.generator,
                    propertyName: PROPERTY_NAME,
                    newValue: !this.isClearButtonEnabled
                };
                const command = await CommandFactory.getCommandFor(this.filterBar, 'Property', modifiedValue, null, flexSettings);
                this.isClearButtonEnabled = !this.isClearButtonEnabled;
                return [command];
            }
            return [];
        }
    }
    var __exports = { __esModule: true };
    __exports.ENABLE_CLEAR_FILTER_BAR_TYPE = ENABLE_CLEAR_FILTER_BAR_TYPE;
    __exports.ToggleClearFilterBarQuickAction = ToggleClearFilterBarQuickAction;
    return __exports;
});
